File: launchpad_zcode/engine.py
"""
Animation Engine and Task State Grid Manager for Launchpad Mini MK3.
Renders task states on the 8x8 matrix and performs visual animations on operations.
"""

import time
import threading
import logging
import random
from typing import Dict, Optional, List, Tuple
from .midi import LaunchpadMiniMK3
from .config import (
    COLOR_OFF,
    COLOR_GREY_DIM,
    COLOR_YELLOW_BRIGHT,
    COLOR_YELLOW_DIM,
    COLOR_GREEN_BRIGHT,
    COLOR_RED_BRIGHT,
    COLOR_CYAN,
    COLOR_BLUE,
    COLOR_WHITE,
    COLOR_MAGENTA,
    COLOR_PURPLE,
    COLOR_ORANGE,
    TASK_COLOR_PENDING,
    TASK_COLOR_ACTIVE,
    TASK_COLOR_SUCCESS,
    TASK_COLOR_ERROR,
)

logger = logging.getLogger(__name__)


class TaskState:
    PENDING = "pending"
    ACTIVE = "active"
    SUCCESS = "success"
    ERROR = "error"


class LaunchpadEngine:
    """
    Manages task grid state (8x8 pads mapped to tasks 0..63)
    and executes temporary visual animations for active operations.
    """

    def __init__(self, launchpad: LaunchpadMiniMK3):
        self.lp = launchpad
        # Map task_id (0..63) -> state ('pending', 'active', 'success', 'error')
        self.tasks: Dict[int, str] = {i: TaskState.PENDING for i in range(64)}
        self._lock = threading.Lock()
        self._animating = False
        self._running = True

        # Start background pulse thread for active tasks
        self._pulse_thread = threading.Thread(target=self._pulse_loop, daemon=True)
        self._pulse_thread.start()

    def stop(self):
        self._running = False

    def task_id_to_grid(self, task_id: int) -> Tuple[int, int]:
        """Convert task ID (0..63) to 8x8 grid (row, col)."""
        task_id = task_id % 64
        row = task_id // 8
        col = task_id % 8
        return (row, col)

    def set_task_state(self, task_id: int, state: str):
        """Update a task's state."""
        with self._lock:
            if 0 <= task_id < 64:
                self.tasks[task_id] = state
        self.render_grid()

    def get_color_for_state(self, state: str, pulse_bright: bool = True) -> int:
        if state == TaskState.PENDING:
            return TASK_COLOR_PENDING
        elif state == TaskState.ACTIVE:
            return TASK_COLOR_ACTIVE if pulse_bright else COLOR_YELLOW_DIM
        elif state == TaskState.SUCCESS:
            return TASK_COLOR_SUCCESS
        elif state == TaskState.ERROR:
            return TASK_COLOR_ERROR
        return COLOR_OFF

    def render_grid(self, pulse_bright: bool = True):
        """Render main task matrix state to Launchpad."""
        if self._animating:
            return

        with self._lock:
            for task_id, state in self.tasks.items():
                row, col = self.task_id_to_grid(task_id)
                color = self.get_color_for_state(state, pulse_bright)
                self.lp.set_grid_pad(row, col, color)

    def _pulse_loop(self):
        """Background loop to pulse active tasks."""
        toggle = False
        while self._running:
            has_active = any(s == TaskState.ACTIVE for s in self.tasks.values())
            if has_active and not self._animating:
                toggle = not toggle
                self.render_grid(pulse_bright=toggle)
            time.sleep(0.5)

    def animate_operation(self, anim_type: str = "spinner", duration: float = 0.8):
        """
        Trigger an operation animation on the Launchpad matrix.
        Supported types:
        'spinner', 'scan', 'success_ripple', 'error_flash',
        'rainbow_wave', 'matrix_rain', 'fireworks', 'snake', 'pulse_beacon'
        """
        def _run_anim():
            self._animating = True
            try:
                if anim_type == "spinner":
                    self._anim_spinner(duration)
                elif anim_type == "scan":
                    self._anim_scan(duration)
                elif anim_type == "success_ripple":
                    self._anim_ripple(COLOR_GREEN_BRIGHT, duration)
                elif anim_type == "error_flash":
                    self._anim_flash(COLOR_RED_BRIGHT, duration)
                elif anim_type == "rainbow_wave":
                    self._anim_rainbow_wave(duration)
                elif anim_type == "matrix_rain":
                    self._anim_matrix_rain(duration)
                elif anim_type == "fireworks":
                    self._anim_fireworks(duration)
                elif anim_type == "snake":
                    self._anim_snake(duration)
                elif anim_type == "pulse_beacon":
                    self._anim_pulse_beacon(duration)
                else:
                    self._anim_spinner(duration)
            finally:
                self._animating = False
                self.render_grid()

        anim_thread = threading.Thread(target=_run_anim, daemon=True)
        anim_thread.start()

    def _anim_spinner(self, duration: float):
        """Rotating spinner around inner 4x4 matrix."""
        coords = [
            (2, 2), (2, 3), (2, 4), (2, 5),
            (3, 5), (4, 5), (5, 5),
            (5, 4), (5, 3), (5, 2),
            (4, 2), (3, 2)
        ]
        start_time = time.time()
        idx = 0
        while time.time() - start_time < duration:
            self.lp.clear()
            r, c = coords[idx % len(coords)]
            self.lp.set_grid_pad(r, c, COLOR_CYAN)
            idx += 1
            time.sleep(0.06)

    def _anim_scan(self, duration: float):
        """Scanning line top to bottom."""
        start_time = time.time()
        row = 0
        direction = 1
        while time.time() - start_time < duration:
            self.lp.clear()
            for col in range(8):
                self.lp.set_grid_pad(row, col, COLOR_WHITE)
            row += direction
            if row > 7 or row < 0:
                direction *= -1
                row += direction * 2
            time.sleep(0.08)

    def _anim_ripple(self, color: int, duration: float):
        """Expanding ripple from center."""
        rings = [
            [(3, 3), (3, 4), (4, 3), (4, 4)],
            [(2, 2), (2, 3), (2, 4), (2, 5), (3, 2), (3, 5), (4, 2), (4, 5), (5, 2), (5, 3), (5, 4), (5, 5)],
            [(1, 1), (1, 6), (6, 1), (6, 6)],
            [(0, 0), (0, 7), (7, 0), (7, 7)]
        ]
        step_delay = duration / len(rings)
        for ring in rings:
            self.lp.clear()
            for r, c in ring:
                self.lp.set_grid_pad(r, c, color)
            time.sleep(step_delay)

    def _anim_flash(self, color: int, duration: float):
        """Flash entire grid."""
        flash_count = 3
        delay = duration / (flash_count * 2)
        for _ in range(flash_count):
            for r in range(8):
                for c in range(8):
                    self.lp.set_grid_pad(r, c, color)
            time.sleep(delay)
            self.lp.clear()
            time.sleep(delay)

    def _anim_rainbow_wave(self, duration: float):
        """Diagonal rainbow wave across the matrix."""
        colors = [COLOR_RED_BRIGHT, COLOR_ORANGE, COLOR_YELLOW_BRIGHT, COLOR_GREEN_BRIGHT, COLOR_CYAN, COLOR_BLUE, COLOR_MAGENTA, COLOR_PURPLE]
        start_time = time.time()
        shift = 0
        while time.time() - start_time < duration:
            self.lp.clear()
            for r in range(8):
                for c in range(8):
                    color_idx = (r + c + shift) % len(colors)
                    self.lp.set_grid_pad(r, c, colors[color_idx])
            shift += 1
            time.sleep(0.08)

    def _anim_matrix_rain(self, duration: float):
        """Matrix digital rain effect with green falling drops."""
        drops = [random.randint(-8, 0) for _ in range(8)]
        start_time = time.time()
        while time.time() - start_time < duration:
            self.lp.clear()
            for col in range(8):
                head = drops[col]
                if 0 <= head < 8:
                    self.lp.set_grid_pad(head, col, COLOR_WHITE)  # Leading edge
                if 0 <= head - 1 < 8:
                    self.lp.set_grid_pad(head - 1, col, COLOR_GREEN_BRIGHT)
                if 0 <= head - 2 < 8:
                    self.lp.set_grid_pad(head - 2, col, COLOR_GREY_DIM)

                drops[col] += 1
                if drops[col] > 10:
                    drops[col] = random.randint(-4, 0)
            time.sleep(0.08)

    def _anim_fireworks(self, duration: float):
        """Random bursting colorful fireworks."""
        start_time = time.time()
        colors = [COLOR_CYAN, COLOR_MAGENTA, COLOR_YELLOW_BRIGHT, COLOR_GREEN_BRIGHT, COLOR_ORANGE]
        while time.time() - start_time < duration:
            center_r = random.randint(1, 6)
            center_c = random.randint(1, 6)
            color = random.choice(colors)

            self.lp.clear()
            self.lp.set_grid_pad(center_r, center_c, COLOR_WHITE)
            time.sleep(0.05)

            # Burst
            for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (-1, 1), (1, -1), (1, 1)]:
                nr, nc = center_r + dr, center_c + dc
                if 0 <= nr < 8 and 0 <= nc < 8:
                    self.lp.set_grid_pad(nr, nc, color)
            time.sleep(0.12)

    def _anim_snake(self, duration: float):
        """Spiral snake tracing the 8x8 matrix boundary inwards."""
        path = []
        top, bottom, left, right = 0, 7, 0, 7
        while top <= bottom and left <= right:
            for c in range(left, right + 1):
                path.append((top, c))
            top += 1
            for r in range(top, bottom + 1):
                path.append((r, right))
            right -= 1
            if top <= bottom:
                for c in range(right, left - 1, -1):
                    path.append((bottom, c))
                bottom -= 1
            if left <= right:
                for r in range(bottom, top - 1, -1):
                    path.append((r, left))
                left += 1

        start_time = time.time()
        idx = 0
        while time.time() - start_time < duration:
            self.lp.clear()
            for i in range(5):  # Snake length 5
                p_idx = (idx - i) % len(path)
                r, c = path[p_idx]
                color = COLOR_MAGENTA if i == 0 else COLOR_PURPLE
                self.lp.set_grid_pad(r, c, color)
            idx += 1
            time.sleep(0.04)

    def _anim_pulse_beacon(self, duration: float):
        """Beacons/waves pulsing from the 4 corners simultaneously."""
        corners = [(0, 0), (0, 7), (7, 0), (7, 7)]
        start_time = time.time()
        step = 0
        while time.time() - start_time < duration:
            self.lp.clear()
            color = COLOR_CYAN if step % 2 == 0 else COLOR_BLUE
            for cr, cc in corners:
                for dr in range(2):
                    for dc in range(2):
                        r = cr + dr if cr == 0 else cr - dr
                        c = cc + dc if cc == 0 else cc - dc
                        self.lp.set_grid_pad(r, c, color)
            step += 1
            time.sleep(0.1)

File: launchpad_zcode_rs/src/engine.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::{Duration, Instant};
use rand::Rng;

use crate::midi::LaunchpadMiniMK3;
use crate::config::*;

#[derive(Clone, Debug, PartialEq)]
pub enum TaskState {
    Pending,
    Active,
    Success,
    Error,
}

pub struct LaunchpadEngine {
    lp: Arc<Mutex<LaunchpadMiniMK3>>,
    tasks: Arc<Mutex<HashMap<usize, TaskState>>>,
    animating: Arc<Mutex<bool>>,
    running: Arc<Mutex<bool>>,
}

impl LaunchpadEngine {
    pub fn new(lp: LaunchpadMiniMK3) -> Self {
        let mut tasks_map = HashMap::new();
        for i in 0..64 {
            tasks_map.insert(i, TaskState::Pending);
        }

        let engine = Self {
            lp: Arc::new(Mutex::new(lp)),
            tasks: Arc::new(Mutex::new(tasks_map)),
            animating: Arc::new(Mutex::new(false)),
            running: Arc::new(Mutex::new(true)),
        };

        engine.start_pulse_thread();
        engine
    }

    pub fn stop(&self) {
        if let Ok(mut r) = self.running.lock() {
            *r = false;
        }
    }

    pub fn set_task_state(&self, task_id: usize, state: TaskState) {
        if task_id < 64 {
            if let Ok(mut t) = self.tasks.lock() {
                t.insert(task_id, state);
            }
            self.render_grid(true);
        }
    }

    pub fn reset(&self) {
        if let Ok(mut t) = self.tasks.lock() {
            for i in 0..64 {
                t.insert(i, TaskState::Pending);
            }
        }
        self.render_grid(true);
    }

    pub fn get_color_for_state(state: &TaskState, pulse_bright: bool) -> u8 {
        match state {
            TaskState::Pending => TASK_COLOR_PENDING,
            TaskState::Active => if pulse_bright { TASK_COLOR_ACTIVE } else { COLOR_YELLOW_DIM },
            TaskState::Success => TASK_COLOR_SUCCESS,
            TaskState::Error => TASK_COLOR_ERROR,
        }
    }

    pub fn render_grid(&self, pulse_bright: bool) {
        if *self.animating.lock().unwrap() {
            return;
        }

        let tasks_guard = self.tasks.lock().unwrap();
        let mut lp_guard = self.lp.lock().unwrap();

        for (&task_id, state) in tasks_guard.iter() {
            let row = (task_id / 8) as u8;
            let col = (task_id % 8) as u8;
            let color = Self::get_color_for_state(state, pulse_bright);
            lp_guard.set_grid_pad(row, col, color);
        }
    }

    fn start_pulse_thread(&self) {
        let tasks_clone = Arc::clone(&self.tasks);
        let animating_clone = Arc::clone(&self.animating);
        let running_clone = Arc::clone(&self.running);
        let lp_clone = Arc::clone(&self.lp);

        thread::spawn(move || {
            let mut toggle = false;
            while *running_clone.lock().unwrap() {
                let has_active = {
                    let t = tasks_clone.lock().unwrap();
                    t.values().any(|s| *s == TaskState::Active)
                };

                if has_active && !*animating_clone.lock().unwrap() {
                    toggle = !toggle;
                    let tasks_guard = tasks_clone.lock().unwrap();
                    let mut lp_guard = lp_clone.lock().unwrap();
                    for (&task_id, state) in tasks_guard.iter() {
                        let row = (task_id / 8) as u8;
                        let col = (task_id % 8) as u8;
                        let color = Self::get_color_for_state(state, toggle);
                        lp_guard.set_grid_pad(row, col, color);
                    }
                }
                thread::sleep(Duration::from_millis(500));
            }
        });
    }

    pub fn animate_operation(&self, anim_type: &str, transition: &str, duration_secs: f32) {
        let lp_clone = Arc::clone(&self.lp);
        let tasks_clone = Arc::clone(&self.tasks);
        let animating_clone = Arc::clone(&self.animating);

        let anim_str = anim_type.to_string();
        let trans_str = transition.to_string();

        thread::spawn(move || {
            {
                let mut anim_guard = animating_clone.lock().unwrap();
                if *anim_guard {
                    // Animation already running, skip or interrupt
                    return;
                }
                *anim_guard = true;
            }

            // Intro transition
            Self::play_transition_intro(&lp_clone, &trans_str);

            // Main animation
            let start = Instant::now();
            let dur = Duration::from_secs_f32(duration_secs);

            match anim_str.as_str() {
                "spinner" => Self::anim_spinner(&lp_clone, start, dur),
                "scan" => Self::anim_scan(&lp_clone, start, dur),
                "success_ripple" => Self::anim_ripple(&lp_clone, COLOR_GREEN_BRIGHT, dur),
                "error_flash" => Self::anim_flash(&lp_clone, COLOR_RED_BRIGHT, dur),
                "rainbow_wave" => Self::anim_rainbow_wave(&lp_clone, start, dur),
                "matrix_rain" => Self::anim_matrix_rain(&lp_clone, start, dur),
                "fireworks" => Self::anim_fireworks(&lp_clone, start, dur),
                "galaxy_spiral" => Self::anim_galaxy_spiral(&lp_clone, start, dur),
                "plasma_wave" => Self::anim_plasma_wave(&lp_clone, start, dur),
                "equalizer_bars" => Self::anim_equalizer_bars(&lp_clone, start, dur),
                "strobe_pulse" => Self::anim_strobe_pulse(&lp_clone, start, dur),
                _ => Self::anim_spinner(&lp_clone, start, dur),
            }

            // Outro transition back to grid
            Self::play_transition_outro(&lp_clone, &tasks_clone, &trans_str);

            {
                let mut anim_guard = animating_clone.lock().unwrap();
                *anim_guard = false;
            }

            // Re-render grid state
            let tasks_guard = tasks_clone.lock().unwrap();
            let mut lp_guard = lp_clone.lock().unwrap();
            for (&task_id, state) in tasks_guard.iter() {
                let row = (task_id / 8) as u8;
                let col = (task_id % 8) as u8;
                let color = Self::get_color_for_state(state, true);
                lp_guard.set_grid_pad(row, col, color);
            }
        });
    }

    // --- Transition Routines ---

    fn play_transition_intro(lp: &Arc<Mutex<LaunchpadMiniMK3>>, trans: &str) {
        match trans {
            "wipe_right" => {
                for col in 0..8 {
                    {
                        let mut lp_guard = lp.lock().unwrap();
                        for row in 0..8 {
                            lp_guard.set_grid_pad(row, col, COLOR_WHITE);
                        }
                    }
                    thread::sleep(Duration::from_millis(15));
                }
            }
            "zoom_iris" => {
                let rings = [
                    [(0, 0), (0, 7), (7, 0), (7, 7)],
                    [(1, 1), (1, 6), (6, 1), (6, 6)],
                    [(2, 2), (2, 5), (5, 2), (5, 5)],
                    [(3, 3), (3, 4), (4, 3), (4, 4)],
                ];
                for ring in rings {
                    {
                        let mut lp_guard = lp.lock().unwrap();
                        for (r, c) in ring {
                            lp_guard.set_grid_pad(r, c, COLOR_MAGENTA);
                        }
                    }
                    thread::sleep(Duration::from_millis(20));
                }
            }
            "dissolve" | _ => {
                let mut rng = rand::thread_rng();
                for _ in 0..20 {
                    let r = rng.gen_range(0..8);
                    let c = rng.gen_range(0..8);
                    {
                        let mut lp_guard = lp.lock().unwrap();
                        lp_guard.set_grid_pad(r, c, COLOR_CYAN);
                    }
                    thread::sleep(Duration::from_millis(5));
                }
            }
        }
    }

    fn play_transition_outro(lp: &Arc<Mutex<LaunchpadMiniMK3>>, tasks: &Arc<Mutex<HashMap<usize, TaskState>>>, trans: &str) {
        let tasks_guard = tasks.lock().unwrap();

        match trans {
            "wipe_right" => {
                for col in 0..8 {
                    {
                        let mut lp_guard = lp.lock().unwrap();
                        for row in 0..8 {
                            let tid = (row * 8 + col) as usize;
                            let color = Self::get_color_for_state(&tasks_guard[&tid], true);
                            lp_guard.set_grid_pad(row, col, color);
                        }
                    }
                    thread::sleep(Duration::from_millis(15));
                }
            }
            _ => {
                let mut lp_guard = lp.lock().unwrap();
                for (&tid, state) in tasks_guard.iter() {
                    let row = (tid / 8) as u8;
                    let col = (tid % 8) as u8;
                    let color = Self::get_color_for_state(state, true);
                    lp_guard.set_grid_pad(row, col, color);
                }
            }
        }
    }

    // --- Visual Animations ---

    fn anim_spinner(lp: &Arc<Mutex<LaunchpadMiniMK3>>, start: Instant, dur: Duration) {
        let coords = [
            (2, 2), (2, 3), (2, 4), (2, 5),
            (3, 5), (4, 5), (5, 5),
            (5, 4), (5, 3), (5, 2),
            (4, 2), (3, 2),
        ];
        let mut idx = 0;
        while start.elapsed() < dur {
            {
                let mut lp_guard = lp.lock().unwrap();
                lp_guard.clear();
                let (r, c) = coords[idx % coords.len()];
                lp_guard.set_grid_pad(r, c, COLOR_CYAN);
            }
            idx += 1;
            thread::sleep(Duration::from_millis(60));
        }
    }

    fn anim_scan(lp: &Arc<Mutex<LaunchpadMiniMK3>>, start: Instant, dur: Duration) {
        let mut row: i8 = 0;
        let mut dir: i8 = 1;
        while start.elapsed() < dur {
            {
                let mut lp_guard = lp.lock().unwrap();
                lp_guard.clear();
                for col in 0..8 {
                    lp_guard.set_grid_pad(row as u8, col, COLOR_WHITE);
                }
            }
            row += dir;
            if row > 7 || row < 0 {
                dir *= -1;
                row += dir * 2;
            }
            thread::sleep(Duration::from_millis(80));
        }
    }

    fn anim_ripple(lp: &Arc<Mutex<LaunchpadMiniMK3>>, color: u8, dur: Duration) {
        let rings = [
            vec![(3, 3), (3, 4), (4, 3), (4, 4)],
            vec![(2, 2), (2, 3), (2, 4), (2, 5), (3, 2), (3, 5), (4, 2), (4, 5), (5, 2), (5, 3), (5, 4), (5, 5)],
            vec![(1, 1), (1, 6), (6, 1), (6, 6)],
            vec![(0, 0), (0, 7), (7, 0), (7, 7)],
        ];
        let step_delay = dur / rings.len() as u32;
        for ring in rings {
            {
                let mut lp_guard = lp.lock().unwrap();
                lp_guard.clear();
                for (r, c) in ring {
                    lp_guard.set_grid_pad(r, c, color);
                }
            }
            thread::sleep(step_delay);
        }
    }

    fn anim_flash(lp: &Arc<Mutex<LaunchpadMiniMK3>>, color: u8, dur: Duration) {
        let delay = dur / 6;
        for _ in 0..3 {
            {
                let mut lp_guard = lp.lock().unwrap();
                for r in 0..8 {
                    for c in 0..8 {
                        lp_guard.set_grid_pad(r, c, color);
                    }
                }
            }
            thread::sleep(delay);
            {
                let mut lp_guard = lp.lock().unwrap();
                lp_guard.clear();
            }
            thread::sleep(delay);
        }
    }

    fn anim_rainbow_wave(lp: &Arc<Mutex<LaunchpadMiniMK3>>, start: Instant, dur: Duration) {
        let colors = [COLOR_RED_BRIGHT, COLOR_ORANGE, COLOR_YELLOW_BRIGHT, COLOR_GREEN_BRIGHT, COLOR_CYAN, COLOR_BLUE, COLOR_MAGENTA, COLOR_PURPLE];
        let mut shift = 0;
        while start.elapsed() < dur {
            {
                let mut lp_guard = lp.lock().unwrap();
                lp_guard.clear();
                for r in 0..8 {
                    for c in 0..8 {
                        let idx = (r + c + shift) % colors.len();
                        lp_guard.set_grid_pad(r as u8, c as u8, colors[idx]);
                    }
                }
            }
            shift += 1;
            thread::sleep(Duration::from_millis(80));
        }
    }

    fn anim_matrix_rain(lp: &Arc<Mutex<LaunchpadMiniMK3>>, start: Instant, dur: Duration) {
        let mut rng = rand::thread_rng();
        let mut drops: Vec<i8> = (0..8).map(|_| rng.gen_range(-8..0)).collect();
        while start.elapsed() < dur {
            {
                let mut lp_guard = lp.lock().unwrap();
                lp_guard.clear();
                for col in 0..8 {
                    let head = drops[col];
                    if (0..8).contains(&head) {
                        lp_guard.set_grid_pad(head as u8, col as u8, COLOR_WHITE);
                    }
                    if (0..8).contains(&(head - 1)) {
                        lp_guard.set_grid_pad((head - 1) as u8, col as u8, COLOR_GREEN_BRIGHT);
                    }
                    drops[col] += 1;
                    if drops[col] > 10 {
                        drops[col] = rng.gen_range(-4..0);
                    }
                }
            }
            thread::sleep(Duration::from_millis(80));
        }
    }

    fn anim_fireworks(lp: &Arc<Mutex<LaunchpadMiniMK3>>, start: Instant, dur: Duration) {
        let mut rng = rand::thread_rng();
        let colors = [COLOR_CYAN, COLOR_MAGENTA, COLOR_YELLOW_BRIGHT, COLOR_GREEN_BRIGHT, COLOR_ORANGE];
        while start.elapsed() < dur {
            let center_r = rng.gen_range(1..7);
            let center_c = rng.gen_range(1..7);
            let color = colors[rng.gen_range(0..colors.len())];

            {
                let mut lp_guard = lp.lock().unwrap();
                lp_guard.clear();
                lp_guard.set_grid_pad(center_r, center_c, COLOR_WHITE);
            }
            thread::sleep(Duration::from_millis(50));

            {
                let mut lp_guard = lp.lock().unwrap();
                for dr in [-1, 0, 1] {
                    for dc in [-1, 0, 1] {
                        let nr = center_r as i8 + dr;
                        let nc = center_c as i8 + dc;
                        if (0..8).contains(&nr) && (0..8).contains(&nc) {
                            lp_guard.set_grid_pad(nr as u8, nc as u8, color);
                        }
                    }
                }
            }
            thread::sleep(Duration::from_millis(120));
        }
    }

    fn anim_galaxy_spiral(lp: &Arc<Mutex<LaunchpadMiniMK3>>, start: Instant, dur: Duration) {
        let mut angle_offset = 0.0f32;
        let colors = [COLOR_CYAN, COLOR_BLUE, COLOR_MAGENTA, COLOR_PURPLE];
        while start.elapsed() < dur {
            {
                let mut lp_guard = lp.lock().unwrap();
                lp_guard.clear();
                for r in 0..8 {
                    for c in 0..8 {
                        let dx = c as f32 - 3.5;
                        let dy = r as f32 - 3.5;
                        let dist = (dx * dx + dy * dy).sqrt();
                        let angle = dy.atan2(dx);
                        let spiral_val = (dist - angle * 2.0 + angle_offset).sin();
                        if spiral_val > 0.4 {
                            let color = colors[(dist as usize) % colors.len()];
                            lp_guard.set_grid_pad(r as u8, c as u8, color);
                        }
                    }
                }
            }
            angle_offset += 0.4;
            thread::sleep(Duration::from_millis(60));
        }
    }

    fn anim_plasma_wave(lp: &Arc<Mutex<LaunchpadMiniMK3>>, start: Instant, dur: Duration) {
        let mut t = 0.0f32;
        let colors = [COLOR_RED_BRIGHT, COLOR_ORANGE, COLOR_YELLOW_BRIGHT, COLOR_GREEN_BRIGHT, COLOR_CYAN, COLOR_MAGENTA];
        while start.elapsed() < dur {
            {
                let mut lp_guard = lp.lock().unwrap();
                lp_guard.clear();
                for r in 0..8 {
                    for c in 0..8 {
                        let v1 = (c as f32 * 0.8 + t).sin();
                        let v2 = (r as f32 * 0.8 + t * 1.2).sin();
                        let v = (v1 + v2 + 2.0) / 4.0;
                        let color_idx = ((v * (colors.len() - 1) as f32) as usize).min(colors.len() - 1);
                        lp_guard.set_grid_pad(r as u8, c as u8, colors[color_idx]);
                    }
                }
            }
            t += 0.3;
            thread::sleep(Duration::from_millis(60));
        }
    }

    fn anim_equalizer_bars(lp: &Arc<Mutex<LaunchpadMiniMK3>>, start: Instant, dur: Duration) {
        let mut rng = rand::thread_rng();
        while start.elapsed() < dur {
            {
                let mut lp_guard = lp.lock().unwrap();
                lp_guard.clear();
                for col in 0..8 {
                    let height = rng.gen_range(1..=8);
                    for r in 0..height {
                        let row = 7 - r;
                        let color = if r >= 6 { COLOR_RED_BRIGHT } else if r >= 4 { COLOR_YELLOW_BRIGHT } else { COLOR_GREEN_BRIGHT };
                        lp_guard.set_grid_pad(row as u8, col as u8, color);
                    }
                }
            }
            thread::sleep(Duration::from_millis(80));
        }
    }

    fn anim_strobe_pulse(lp: &Arc<Mutex<LaunchpadMiniMK3>>, start: Instant, dur: Duration) {
        let mut step = 0;
        while start.elapsed() < dur {
            {
                let mut lp_guard = lp.lock().unwrap();
                lp_guard.clear();
                for r in 0..8 {
                    for c in 0..8 {
                        if (r + c) % 2 == (step % 2) {
                            lp_guard.set_grid_pad(r as u8, c as u8, COLOR_WHITE);
                        } else {
                            lp_guard.set_grid_pad(r as u8, c as u8, COLOR_BLUE);
                        }
                    }
                }
            }
            step += 1;
            thread::sleep(Duration::from_millis(80));
        }
    }
}
